### PR TITLE
Adding processing of EMG time series into covariance matrices

### PR DESCRIPTION
**Pull request by MariusGuerard**
- Created prepare_emg_data.py to batch the time series data and compute covariance matrices into the batches.
- Display mean spd matrices (euclidean and Affine-invariant) for the different signs

### CODE REVIEW

File: geomstats/datasets/prepare_emg_data.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+"""Pre-process emg time series into batched covariance matrices.
```
Comment (ninamiolane) [inline geomstats/datasets/prepare_emg_data.py:1]: NIT: Maybe just give the acronym of EMG, since most users might not be familiar with it?

File: geomstats/datasets/prepare_emg_data.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+The user defines the number of time steps of the batches.
+"""
+
+import numpy as np
+import pandas as pd
+
+
+def _vectorize_cov(cov_mat):
+    """Vectorize a symmetric Matrix.
```
Comment (ninamiolane) [inline geomstats/datasets/prepare_emg_data.py:11]: Ah! We already have this function here :/ 

https://github.com/geomstats/geomstats/blob/544debcf81dc40d7e5d584c5606792c2cfa9f4eb/geomstats/geometry/symmetric_matrices.py#L81

File: geomstats/datasets/prepare_emg_data.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    lcov = len(cov_mat)
+    covec = []
+    for i in range(lcov):
+        for j in range(i + 1):
+            covec.append(cov_mat[i][j])
+    return np.array(covec)
+
+
+class TimeSerieCovariance:
```
Comment (ninamiolane) [inline geomstats/datasets/prepare_emg_data.py:34]: NIT: one time-series (the singular has the s)

File: geomstats/datasets/prepare_emg_data.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    """Class for generating a list of covariance matrix from time series.
```
Comment (ninamiolane) [inline geomstats/datasets/prepare_emg_data.py:35]: NIT: list of covariance matrices (plural)

File: geomstats/datasets/prepare_emg_data.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    It starts by removing the transient signal by taking a margin on each side
```
Comment (ninamiolane) [inline geomstats/datasets/prepare_emg_data.py:37]: This documentation, and some parts of the code, seem to be specific to your EMG data (since it incorporates sign changes?). Two options:
- Either only keep general code in this class, that could be applied to any time-series. The code specific to your EMG data would go to the ipynb.
- Or make it clear that the preprocessing is meant for your EMG data, and give the link to the data in the docstring.

I would maybe prefer option 1, because it is likely that such general code (creation of batches and covariance matrices from any timeseries) will be interesting to users. But this is your choice!

**Review by ninamiolane — COMMENTED** [review]
First comments

File: geomstats/datasets/prepare_emg_data.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    df : pd.DataFrame
+        Output data containing the covariance matrices along with the labels.
+    """
+
+    def __init__(self, data, n_steps, n_elec):
+        self.label_map = {'rock': 0, 'scissors': 1, 'paper': 2, 'ok': 3}
+        self.data = data
+        self.n_steps = n_steps
+        self.n_elec = n_elec
+        self.batches = np.array([])
```
Comment (ninamiolane) [inline geomstats/datasets/prepare_emg_data.py:76]: To make it work with numpy/autograd/pytorch/tensorflow, use:
```
import geomstats.backend as gs
```
and then:
`gs.array, gs.append`, etc

File: geomstats/datasets/prepare_emg_data.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        self.covs = np.array([])
+        self.df = pd.DataFrame()
```
Comment (ninamiolane) [inline geomstats/datasets/prepare_emg_data.py:78]: To make it more general, let's only use array structures like gs.array (no `pd.Dataframe` within the numerical core, let's just give different names to different variables).

File: geomstats/datasets/prepare_emg_data.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        Adding a time margin at each sign change to get stationary
+        signal corresponding to each sign.
+        """
+        start_sign = np.where(np.diff(self.data.y) != 0)[0]
+        end_sign = np.append(start_sign[1:], len(self.data)) - 1000
+        start_sign += 1000
+        self.batches = np.concatenate([range(start_sign[j],
+                                             end_sign[j] - self.n_steps,
+                                             self.n_steps)
+                                      for j in range(len(start_sign))])
```
Comment (ninamiolane) [inline geomstats/datasets/prepare_emg_data.py:97]: `for start_id, end_id in zip(start_sign_ids, end_sign_ids)` would be more compact and easier to read.

File: geomstats/datasets/prepare_emg_data.py (hunk 7 above)
Comment (ninamiolane) [reply geomstats/datasets/prepare_emg_data.py:78]: Same with the `data` structure.

File: tests/test_prepare_emg_data.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
+"""Unit tests for TimeSerieCovariance class."""
+
+import numpy as np
```
Comment (ninamiolane) [inline tests/test_prepare_emg_data.py:3]: `import geomstats.backend as gs`

**Review by ninamiolane — COMMENTED** [review]
Few more comments, all details!

File: notebooks/07_emg_hand_gesture_classification.ipynb [hunk 10]
```diff
@@ -50,24 +50,6 @@
     "Note: Another interesting use case, not explored in this notebook, would be to test what is the accruacy when calibrating, removing the armband and giving it to someone else that is calibrating it on its own arm before testing it. The idea being that transfer learning might help getting better results (or faster calibration) than calibrating on one user."
```
Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:50]: <p>NIT: Instead of comments, we try to write directly in the text cells of Jupyter: it is slightly more clean and more userfriendly.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/903/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:50]: <p>indexing as <code>ax[i // 2, i % 2]</code> could maybe do the trick and avoid the if condition?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/903/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:50]: <p>NIT: "...does not show a striking visual difference".</p><p> </p><p>Except maybe for Rock that seems distinct?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/903/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:50]: <p>Nice text!</p><p> </p><p>NIT: qre -&gt; are</p><p>invariant metrics -&gt; metric (singular?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/903/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:50]: <p>Explain why we are printing these lines here? (Is it a leftover from tests you were doing?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/903/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

**Comment by ninamiolane** [comment]
Very nice!! 🎉  We should really try to merge #623 soon as it could be useful for this notebook.

I have just a few comments on the organization of the code (separating between general code - that will be re-used by many) and the code that is specific to this EMG notebook.

Also, let's try to use geomstats backend (gs. syntax) so that this code can be run seemlessly on pytorch or tensorflow.

File: notebooks/07_emg_hand_gesture_classification.ipynb (hunk 10 above)
Comment (MariusGuerard) [reply notebooks/07_emg_hand_gesture_classification.ipynb:50]: Oui bonne idée!

**Comment by ninamiolane** [comment]
The Travis tests are failing in pytorch and tensorflow. You can add the decorator `@geomstats.tests.np_only` on top of the function testing your notebook to solve this problem:
https://travis-ci.org/github/geomstats/geomstats/builds/756668204

File: notebooks/07_emg_hand_gesture_classification.ipynb (hunk 10 above)
Comment (ninamiolane) [reply notebooks/07_emg_hand_gesture_classification.ipynb:50]: <p>N_ELECTRODES should be N_ELEC?</p>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:50]: <p>N_ELEC has already been defined, as N_ELECTRODES, above.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/903/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:50]: <p><code>for i in range(N_ELECTRODES)</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/903/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:50]: <p>spd.SPDMatrices(N_ELECTRODES)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/903/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

**Comment by ninamiolane** [comment]
Awesome! Just another set of very minor comments and we are ready to merge  🎉 It's exciting to see this notebook being written, it's definitely a fascinating application of our metrics.

**Comment by ninamiolane** [comment]
@MariusGuerard Travis tests are failing, see: https://travis-ci.org/github/geomstats/geomstats/jobs/757972512

This can be solved by adding the decorator @geomstats.tests.np_only on top of the tests in test_prepare_emg_data 🎉
